Implement Tree Search Functionality with Highlighting and Filtering

monxa opened the pull request:
Closes #192

Tested with GDExtension and Module.

Features:
- *Tree highlighting*: Highlights items that match the search query.
- *Counting descendants*: Shows the number of matching items within collapsed branches.
- *Jump to next match*: on enter.
- (Limbo-)*Shortcut*: Default CTRL-F. 
- *Menu entry*: Misc->Search Tree.

Known issues:
- [x] Unnecessarily redundant signals ([af15dde](https://github.com/limbonaut/limboai/pull/229/commits/af15dde4c896b76db66689d49ced72730b8e7775))
-  [x] ~~`TreeSearch::_draw_highlight_item`: uses one magic number~~ (same as engine code)
- Performance
   - [x] `TreeSearch::_find_matching_entries` et al ([cc8f099](https://github.com/limbonaut/limboai/pull/229/commits/cc8f099d82f9aeaa3c3affd2fb3bf588834bdd5e))
     - Can be solved by making it iterative instead of recursive, using linear push_back instead of insert and sorting once.
   - [x] Searches do a whole tree reconstruction for now: Noticeable lag. (https://github.com/monxa/limboai/pull/1)
- [x] Editor scaling needs fixing ([380f80c](https://github.com/limbonaut/limboai/pull/229/commits/380f80c2b3166548f01743bd96e21e2ef418135d))
- [x] ~~Searchbar spacing should be adjusted.~~ (Already similar to Docs/Script search bar with [8c32395](https://github.com/limbonaut/limboai/pull/229/commits/8c323956e3afe74243fff791ff807167bd913baf))
- [ ] ~~Word highlight drawing might be too obstructive. Especially in filter mode.~~ (Will be solved after this is merged)
- [x] Jumping through entries currently inconsistent ([8d29f16](https://github.com/limbonaut/limboai/pull/229/commits/8d29f16963c30dcd073964b028ff6d5f1303b952), [47706e9](https://github.com/limbonaut/limboai/pull/229/commits/47706e9480b6fc831e171f9a6c39ca0d7b6d988c))
- [x] Filtering unstable. ([74cf985](https://github.com/limbonaut/limboai/pull/229/commits/74cf9855a7fed51ed6e373ed14599a64a62eae46), [465757e](https://github.com/limbonaut/limboai/pull/229/commits/465757eec830a73da446973bcce2c640e9a794a6))
- [ ] Decide if the current case matching behavior needs adjustment. Current: Match case when input includes capital letter.

Wishlist (non-blockers)
1. Word highlight drawing might be too obstructive. Especially in filter mode.
2. - [x] Remember separate ~~search-mask~~ `SearchInfo` for each tab. ([10c8f58](https://github.com/limbonaut/limboai/pull/229/commits/10c8f58ca9b9e3c9034b7eb5b5dc52e2677a18c7))
3. - [x] Buttons for select next/previous entry (align with script/docs viewport). ([ec6c4c3](https://github.com/limbonaut/limboai/pull/229/commits/ec6c4c3036a62a9d8f47f4f61f610cff95964389))
4. Remember and restore tree collapse on jumping

This pr includes a significant amount of new code (~500 lines). I would greatly appreciate any feedback, especially regarding the known issues listed above.

[demo_filter_tree.webm](https://github.com/user-attachments/assets/76fb8ed1-3ff2-4fa7-948a-101ed633b7ac)

monxa commented on editor/tree_search.cpp:
```diff
@@ -0,0 +1,481 @@
+}
+
+// custom draw callback for highlighting (bind 2)
+void TreeSearch::_draw_highlight_item(TreeItem *p_tree_item, Rect2 p_rect, String p_search_mask, Callable p_parent_draw_method) {
```
Probably better to not bind `search_mask`.

monxa replied:
[ffe344d](https://github.com/limbonaut/limboai/pull/229/commits/ffe344d16651307c7c44e1f29897c666a516006d) addresses this.

monxa commented:
Jumping through entries currently behaves weirdly: Selection does not always get represented correctly in the tree view. 
I think that has to do with unfolding through code and instantly after changing the selection. Will look into this.

limbonaut commented:
Thanks for submitting! Since it's a big chunk of code, I need some time to review it fully. 

Got a crash during testing: with an active search, I switched to another tab and pressed enter. It's one of those cached `TreeItem` pointers, judging by the backtrace. 

Thinking about it, search state should probably be remembered per tab. I only skimmed through your code a little bit, and it doesn't look like it's the case, but would require changes to achieve that. Maybe `TreeSearch` instances could be kept per tab? But those `TreeItem` pointers are a problem then - they must be invalidated as soon as the user switches to another tab (or Tree is rebuilt).

monxa commented on editor/tree_search.h:
```diff
@@ -0,0 +1,121 @@
+	// we will add everything from TaskTree.h
+	void update_search(Tree *p_tree);
+	void on_item_edited(TreeItem *p_item);
+	TreeSearchPanel *search_panel;
```
So you are proposing to construct the `TreeSearchPanel` in `TaskTree` and calling `TreeSearch(tree_search_panel)`?

I wouldn't even though of that. Connecting the `update_requested_signal` causes no issue then, because we have access.

The TaskTree would then maintain privately maintain TreeSearchPanel * tree_search_panel for 
```
void TaskTree::tree_search_show_and_focus() {
	tree_search->search_panel->show_and_focus();
}
```
(which is called by the misc entry).
Correct?

limbonaut replied:
It could be owned either by the TaskTree or by the Editor, whatever makes more sense. Haven't read the code in fullest yet.

monxa replied:
[47706e9](https://github.com/limbonaut/limboai/pull/229/commits/47706e9480b6fc831e171f9a6c39ca0d7b6d988c) Should address this.

monxa commented:
> Thanks for submitting! Since it's a big chunk of code, I need some time to review it fully.
> 
> Got a crash during testing: with an active search, I switched to another tab and pressed enter. It's one of those cached `TreeItem` pointers, judging by the backtrace.
> 
> Thinking about it, search state should probably be remembered per tab. I only skimmed through your code a little bit, and it doesn't look like it's the case, but would require changes to achieve that. Maybe `TreeSearch` instances could be kept per tab? But those `TreeItem` pointers are a problem then - they must be invalidated as soon as the user switches to another tab (or Tree is rebuilt).

Great catch!

monxa commented:
Yes this is a good chunk of code. And I expect much of it not to align with what you want. Please take your time for review(s).

Will later find some time to address invalidation and add some FAIL-conds.

You are thinking into an interesting direction with your proposal of caching multiple tabs. 
I think that would need the TreeSearch to have access to TaskTree. I tried to avoid that in favor of molecularity and to avoid circular includes. If you need this to be changed into such an direction tho: I am very open to make the necessary changes.

monxa commented:
> Got a crash during testing: with an active search, I switched to another tab and pressed enter. It's one of those cached TreeItem pointers, judging by the backtrace.

[47706e9](https://github.com/limbonaut/limboai/pull/229/commits/47706e9480b6fc831e171f9a6c39ca0d7b6d988c)
Fixed a rather prevalent issue within `TreeSearch::_select_item` that led to the nullptr exception.

[9a1641e](https://github.com/limbonaut/limboai/pull/229/commits/9a1641e8ab608ce4d0c94ce7dd108ba128351eed)
Moved the TreeSearch::update_tree call from `TaskTree::_update_tree` to `TaskTree::_create_tree`. This should catch all cases where the task tree gets updated, also those where it gets updated from the Editor Plugin directly. AFAIK, the caching approach is infeasible. We don't cache trees at all after all. If you want to transition towards such a solution in the future, I apologize for any additional implementation considerations this solution causes.

limbonaut commented on editor/tree_search.cpp:
```diff
@@ -0,0 +1,492 @@
+}
+
+Vector<TreeItem *> TreeSearch::_find_matching_entries(TreeItem *p_tree_item, const String &p_search_mask, Vector<TreeItem *> &p_accum) {
+	if (!p_tree_item)
```
Please use {} for consistency and same below.

monxa replied:
Sorry.

limbonaut commented on editor/tree_search.cpp:
```diff
@@ -0,0 +1,492 @@
+	return search_panel->get_text();
+}
+
+Vector<TreeItem *> TreeSearch::_find_matching_entries(TreeItem *p_tree_item, const String &p_search_mask, Vector<TreeItem *> &p_accum) {
```
Argument name differs in the header. If it's a return argument, it should be called `r_accum` or `r_buffer` - this is how it's done in other places. There seems to be no need to explicitly return it, either.

monxa replied:
Inconsistency resolved. Is now void.

limbonaut commented on editor/tree_search.cpp:
```diff
@@ -0,0 +1,492 @@
+		if (!_vector_has_bsearch(matching_entries, item)) {
+			continue;
+		}
+		String debug_string = "[";
```
Unused variable

monxa replied:
Done.

limbonaut commented on editor/tree_search.cpp:
```diff
@@ -0,0 +1,492 @@
+	}
+
+	// find the best fitting entry.
+	for (int i = 0; i < ordered_tree_items.size(); i++) {
```
Suggestion: Start a loop from `MAX(0, selected_idx)` and drop the second part of the condition below

monxa replied:
new version:
```cpp
// find the best fitting entry.
for (int i = MAX(0, selected_idx) + 1; i < ordered_tree_items.size(); i++) {
  TreeItem *item = ordered_tree_items[i];
  if (_vector_has_bsearch(matching_entries, item)) {
	  _select_item(item);
	  return;
}
}
```

limbonaut commented on editor/tree_search.cpp:
```diff
@@ -0,0 +1,492 @@
+	_update_number_matches();
+
+	_highlight_tree(search_mask);
+	if (!search_panel->is_connected("text_submitted", callable_mp(this, &TreeSearch::_select_next_match))) {
```
Should be in `NOTIFICATION_READY`

monxa replied:
Since this is `RefCounted`, I now put it into the constructor. is_connected needs no checking this way. Great catch!

monxa replied:
Done.

limbonaut reviewed (changes requested):
Beware, nitpicks incoming :rofl:

limbonaut commented:
Some additional suggestions:
- Show "Ctrl+F" shortcut in the Misc menu (already mentioned in a comment)
- Remember search state per UI tab (borderline out-of-scope - we can make an issue to track it instead)
Also, case sensitivity can be unintuitive: I mistyped "playAnimation" and was surprised that it didn't find anything. 
I've tested it, and it works really well - didn't find any other issues. If you are concerned about the selection box being confusing, we can try a custom StyleBox: a filled square built from the font color with the alpha set somewhere in [0.1,0.9] range - should work with any theme, I believe.

limbonaut commented:
I've tested it, and it works really well - didn't find any other issues. If you are concerned about the selection box being confusing, we can try a custom StyleBox: a filled square built from the font color with the alpha set somewhere in [0.1,0.9] range - should work with any theme, I believe.

Some additional suggestions:
- Show "Ctrl+F" shortcut in the Misc menu (already mentioned in a comment)
- Remember search state per UI tab (borderline out-of-scope - we can make an issue to track it instead)
- Also, case sensitivity can be unintuitive: I mistyped "playAnimation" and was surprised that it didn't find anything (not sure if it's an issue, though).

monxa commented on editor/task_tree.cpp:
```diff
@@ -105,6 +112,7 @@ void TaskTree::_update_item(TreeItem *p_item) {
 	if (!warning_text.is_empty()) {
 		p_item->add_button(0, theme_cache.task_warning_icon, 0, false, warning_text);
 	}
+	tree_search->on_item_edited(p_item); // this is necessary to preserve custom drawing from tree search.
```
Great suggestion! This really improves the code.

monxa commented on editor/tree_search.h:
```diff
@@ -0,0 +1,126 @@
+
+using namespace godot;
+
+enum TreeSearchMode {
```
Since we want to eventually separate the TreeSearchPanel from TreeSearch, I propose making TreeSearchMode a member of TreeSearch.

That requires TreeSearch to be predecessor to TreeSearchPanel in the header while forward declaring the TreeSearchPanel. Done.

monxa commented on editor/tree_search.cpp:
```diff
@@ -0,0 +1,492 @@
+
+/* ------- TreeSearchPanel ------- */
+
+void TreeSearchPanel::_initialize_controls() {
```
Ah there is one reason for this: We can't get the theme icon at constructor time. New solution: Moved `_initialize_controls`-call into constructor. BUTTON_SET_ICON now is the only remaining part of `_initialize_controls` in `NOTIFICATION_READY`. Does this suffice?

monxa replied:
It's in `NOTIFICATION_THEME_CHANGED` now.

monxa commented on editor/tree_search.cpp:
```diff
@@ -0,0 +1,492 @@
+}
+
+void TreeSearchPanel::_emit_text_submitted(const String &p_text) {
+	this->emit_signal("text_submitted");
```
Got it. All `this->`-syntax removed.

monxa commented on editor/tree_search.cpp:
```diff
@@ -0,0 +1,492 @@
+void TreeSearchPanel::_notification(int p_what) {
+	switch (p_what) {
+		case NOTIFICATION_READY: {
+			_initialize_controls();
```
See https://github.com/limbonaut/limboai/pull/229#discussion_r1779665459

monxa replied:
Done

monxa commented on editor/tree_search.cpp:
```diff
@@ -0,0 +1,492 @@
+}
+
+String TreeSearchPanel::get_text() {
+	if (!line_edit_search)
```
Sorry.

monxa replied:
I think all cases of this are caught now.

limbonaut replied:
No worries. I'm trying to keep code with the same conventions as Godot, in most cases.

monxa replied:
I like that!

monxa commented on editor/tree_search.h:
```diff
@@ -0,0 +1,126 @@
+	FILTER = 1
+};
+
+class TreeSearchPanel : public HFlowContainer {
```
No! We had a HBoxContainer before [f523af2](https://github.com/limbonaut/limboai/pull/229/commits/f523af2d13bea14e09ba8dcae92c09ae5f67505c).
HFlowContainer allows for smaller minimum sizes. Otherwise we'd need something like a ScrollContainer in-between. I tried that. Did increase minimum v-size. Didn't like that. What do you think?

limbonaut replied:
HBox's minimum is the combined minimum of its children, and the smallest size it can have is zero size. I don't think I understand what you mean. I switched it to HBoxContainer - and took a screenshot - identical result. Am I missing something?

monxa replied:
current - `HFlowContainer`
[demo_hflowcontainer.webm](https://github.com/user-attachments/assets/f032176b-5619-4865-a8cc-3ee89151d4b1)

[a80b737](https://github.com/limbonaut/limboai/pull/229/commits/a80b737319da1ffeead4c3e250f303d6844817ae) `HBoxContainer`
[demo_hboxcontainer.webm](https://github.com/user-attachments/assets/2f588f16-4b50-4970-92af-c3ec70131dda)

limbonaut replied:
Oh wow, okay, that's super trippy. If it works better with HFlow, let's keep that, then.

monxa commented on editor/tree_search.cpp:
```diff
@@ -0,0 +1,492 @@
+}
+
+bool TreeSearchPanel::has_focus() {
+	return false;
```
Haha. This is indeed very funny. I think I just introduced this template. Will remove for now.

monxa commented on editor/tree_search.cpp:
```diff
@@ -0,0 +1,492 @@
+		Vector2 substring_before_size = font->get_string_size(substring_before, HORIZONTAL_ALIGNMENT_LEFT, -1.f, font_size);
+
+		// stylebox
+		Ref<StyleBox> stylebox = p_tree_item->get_tree()->get_theme_stylebox("Focus");
```
Sure. We can do that!

monxa commented on editor/tree_search.cpp:
```diff
@@ -0,0 +1,492 @@
+	switch (p_what) {
+		case NOTIFICATION_READY: {
+			_initialize_controls();
+			line_edit_search->connect("text_changed", callable_mp(this, &TreeSearchPanel::_emit_update_requested).unbind(1));
```
Thank you very much!

monxa replied:
Nice. Thanks!

monxa replied:
This needs a tiny bit of alteration. We can't bind internal callables using this MethodInfo macro.

monxa replied:
This okay?
```
// search callbacks
Callable c_update_requested = Callable(this, "emit_signal").bind("update_requested");
Callable c_text_submitted = Callable((Object*)this, "emit_signal").bind("text_submitted");

line_edit_search->connect("text_changed", c_update_requested.unbind(1));
check_button_filter_highlight->connect("pressed", c_update_requested);
line_edit_search->connect("text_submitted", c_text_submitted.unbind(1));
```

limbonaut replied:
> This needs a tiny bit of alteration. We can't bind internal callables using this MethodInfo macro.

Worked in the past, hmm. The issue is probably with the varargs - can't bind to member functions with variable arguments? Binding to Resource::emit_changed works, but that function takes zero arguments. 
Both a Callable or a dedicated method is okay, then, whichever you prefer.

monxa replied:
The above is already pushed.

monxa commented on editor/task_tree.cpp:
```diff
@@ -562,12 +573,24 @@ void TaskTree::_bind_methods() {
 	editable = true;
 	updating_tree = false;
 
+	// for Tree + TreeSearch, we want a VBoxContainer. For now, rather than changing this classes type, let's do nesting:
```
Addressed.

monxa commented on editor/tree_search.cpp:
```diff
@@ -0,0 +1,492 @@
+	// Recursively collect items from the first child
+	TreeItem *child = p_tree_item->get_first_child();
+	while (child) {
+		_update_ordered_tree_items(child);
```
Partly actually addressed. Remaining consideration: Updating the whole tree whenever `search_mask`

monxa commented on editor/limbo_ai_editor_plugin.cpp:
```diff
@@ -1319,6 +1324,9 @@ void LimboAIEditor::_update_misc_menu() {
 			MISC_CREATE_SCRIPT_TEMPLATE);
+	
+	misc_menu->add_separator();
+	misc_menu->add_icon_item(theme_cache.search_icon, ("Search Tree"), MISC_SEARCH_TREE);
```
Addressed.

monxa commented on editor/limbo_ai_editor_plugin.cpp:
```diff
@@ -1512,6 +1521,9 @@ LimboAIEditor::LimboAIEditor() {
 	LW_SHORTCUT("limbo_ai/jump_to_owner", TTR("Jump to Owner"), (Key)(LW_KEY_MASK(CMD_OR_CTRL) | LW_KEY(J)));
 	LW_SHORTCUT("limbo_ai/close_tab", TTR("Close Tab"), (Key)(LW_KEY_MASK(CMD_OR_CTRL) | LW_KEY(W)));
+	LW_SHORTCUT("limbo_ai/find_task", TTR("Find Task"), (Key)(LW_KEY_MASK(CMD_OR_CTRL) | LW_KEY(F)));
+	LW_SHORTCUT("limbo_ai/hide_tree_search", TTR("Hide BehaviorTrees Search Panel"), (Key)(LW_KEY(ESCAPE)));
```
Addressed.

monxa commented on editor/tree_search.cpp:
```diff
@@ -0,0 +1,492 @@
+	_add_spacer(0.25);
+}
+
+void TreeSearchPanel::_initialize_close_callbacks() {
```
Done.

limbonaut commented on editor/tree_search.cpp:
```diff
@@ -95,6 +93,9 @@ void TreeSearchPanel::_notification(int p_what) {
 		}
+		case NOTIFICATION_THEME_CHANGED: {
+			label_filter->set_text(TTR("Filter"));
+		}
```
Missing `break`. In Godot, there is a convention to use such idiom:
```cpp
case X: {
} break;
```
I think we should adopt it as well.

monxa replied:
Done.

monxa commented:
I think I am actually being a bit to cautious about [d956551](https://github.com/monxa/limboai/commit/d956551dd4178744eafb3cd5990a20cfee5ab45b) and that can be merged into this. Will look into it tomorrow.

monxa commented:
I think I am actually a bit too cautious about [d956551](https://github.com/monxa/limboai/commit/d956551dd4178744eafb3cd5990a20cfee5ab45b). Will merge that into this tomorrow.

monxa commented on editor/tree_search.h:
```diff
@@ -134,7 +140,10 @@ class TreeSearchPanel : public HFlowContainer {
-	void show_and_focus();
+	TreeSearch::SearchInfo get_search_info() const;
+	void set_search_info(TreeSearch::SearchInfo p_search_info);
+	void set_text(const String &p_text);
```
`set_text` can be dropped in favor of `set_search_info`.

monxa commented on editor/tree_search.cpp:
```diff
@@ -513,8 +513,8 @@ void TreeSearchPanel::_notification(int p_what) {
 
 void TreeSearchPanel::_bind_methods() {
 	ADD_SIGNAL(MethodInfo("update_requested"));
```
I am not sure whether we should also replace this with a LW_NAME entry.

monxa replied:
This is the only remaining hard-coded c-string.

monxa replied:
(For signals that is)

limbonaut replied:
Generally, in constructors and in `_bind_method`, we are not doing that. Unless in classes that are often created. It makes the most sense in places that are repeatedly executed.

monxa replied:
Do you want me to revert this?

limbonaut replied:
Not really, no.

monxa commented on editor/tree_search.cpp:
```diff
@@ -68,10 +68,16 @@ void TreeSearch::_filter_tree(const String &p_search_mask) {
+		bool parents_visible = true;
 		while (first_counting_ancestor && !number_matches.has(first_counting_ancestor)) {
+			// Performance: we only need to check the first visible ancestor. This is already optimal because of the ordering in ordered_tree_items.
+			if (!first_counting_ancestor->is_visible()){
```
This could sit happily in the while-condition. But I *feel* like the additional is_visible api-call does introduce minuscule, tho noticeable, amount of lag.

monxa replied:
I don't feel *any* lag filtering with this solution. Tho, if this does have a greater impact, we could also track seen items by explicitly marking their pointers in a local `HashMap`.

monxa replied:
I checked the godot source. TreeItem::is_visible simply returns visible. My *feeling* was probably unfounded. The api calls through GDExtension should not matter in the greater context. Still, the this solution not good.

monxa commented on editor/tree_search.cpp:
```diff
@@ -56,29 +56,41 @@ void TreeSearch::_clean_callable_cache() {
-	for (int i = 0; i < ordered_tree_items.size(); i++) {
-		TreeItem *cur_item = ordered_tree_items[i];
+	Vector<bool> item_visibilities;
+	item_visibilities.resize(ordered_tree_items.size());
```
This is not 0-initialized! We simply get some memory. Fill with false!

monxa commented on editor/task_tree.cpp:
```diff
@@ -17,22 +17,23 @@
-#include <godot_cpp/classes/texture_rect.hpp>
 #include <godot_cpp/classes/label.hpp>
+#include <godot_cpp/classes/script.hpp>
+#include <godot_cpp/classes/texture_rect.hpp>
```
I'd like that. I especially would like to see the commit so I could use similar workflows myself.

monxa commented:
As discussed. Squash incoming...

ydeltastar commented:
It works really well even with drag and drop. The only major issue I noticed is that when I have a task named `AnimationTreeTravel`, "animationtreetravel" gets a match but "animationTreetravel" or other case variations don't.

monxa commented:
> It works really well even with drag and drop. The only major issue I noticed is that when I have a task named `AnimationTreeTravel`, "animationtreetravel" gets a match but "animationTreetravel" or other case variations don't.

Yes. The current behavior: Match case when input contains capital letter. Might need adjustment: Not intuitive. @limbonaut also pointed this out. 

IMO, we can easily change that after this is merged.
We can do either:
- Always case-insensitive (Same as in docs)
- "Match Case"-Checkbox (Same as in script editor)

limbonaut commented:
> It works really well even with drag and drop. The only major issue I noticed is that when I have a task named `AnimationTreeTravel`, "animationtreetravel" gets a match but "animationTreetravel" or other case variations don't.

From reading the code, I think the idea is that if you supply a lower-case string, the search assumes case-insensitive, and if you mix-in CAPS - it assumes case-sensitive. We are trying to establish if it's a bug or a feature.

ydeltastar commented:
> Match case when input contains capital letter.

The search's placeholder text should probably include this as a hint.

limbonaut commented on editor/tree_search.cpp:
```diff
@@ -0,0 +1,654 @@
+}
+
+TreeSearchPanel::TreeSearchPanel() {
+	_initialize_controls();
```
Could you move `_initialize_controls` code straight into constructor? It doesn't look like it needs to be a separate function.

monxa replied:
Inlined.

limbonaut commented on editor/tree_search.cpp:
```diff
@@ -0,0 +1,654 @@
+}
+
+String TreeSearchPanel::get_text() const {
+	if (!line_edit_search) {
```
`line_edit_search` is allocated in the constructor, so this check is unnecessary

monxa replied:
Removed.

limbonaut reviewed (changes requested):
Looking good! I've added just a few suggestions. In testing, it works like a charm. 

We can consider if the case-sensitive issue needs changing, but no need to block this PR until that decision is made.

monxa commented on editor/limbo_ai_editor_plugin.cpp:
```diff
@@ -280,6 +284,18 @@ void LimboAIEditor::edit_bt(const Ref<BehaviorTree> &p_behavior_tree, bool p_for
 
+	// Restore search info from [tab_search_context].
+	if (idx_history >= 0 && idx_history < history.size()) {
+		// info for BehaviorTree available. Restore!
```
Done.

monxa commented on editor/limbo_ai_editor_plugin.h:
```diff
@@ -134,12 +136,14 @@ class LimboAIEditor : public Control {
 	EditorLayout editor_layout;
 	Vector<Ref<BehaviorTree>> history;
 	int idx_history;
+	HashMap<Ref<BehaviorTree>, TreeSearch::SearchInfo> tab_search_context;
```
Now removing `<BehaviorTree, SearchInfo>` reference in `_tab_closed`. Also fixed `SearchInfo` restoration on tab closed.

monxa commented on editor/limbo_ai_editor_plugin.cpp:
```diff
@@ -799,6 +817,9 @@ void LimboAIEditor::_misc_option_selected(int p_id) {
 		} break;
+		case MISC_SEARCH_TREE: {
+			task_tree->tree_search_show_and_focus();
+		}
```
Sorry. Done.

monxa commented on editor/tree_search.h:
```diff
@@ -0,0 +1,161 @@
+	void _highlight_tree_item(TreeItem *p_tree_item);
+
+	// Custom draw-Callback (bind inherited Callable).
+	void _draw_highlight_item(TreeItem *p_tree_item, const Rect2 p_rect, const Callable p_parent_draw_method);
```
Yup. Works. Done.

monxa commented:
[MacOS task](https://github.com/limbonaut/limboai/blob/master/.github/workflows/macos.yml) appears to be broken.

monxa commented:
> Match case when input contains capital letter.

Does a tooltip for the SearchBox suffice for now?
```cpp
line_edit_search->set_tooltip_text("Match case if input contains capital letter.");
```

limbonaut commented:
> [MacOS task](https://github.com/limbonaut/limboai/blob/master/.github/workflows/macos.yml) appears to be broken.
> 
> Fixed on master. See: [godotengine/godot#97981](https://github.com/godotengine/godot/pull/97981)

Hmm, bummer.

limbonaut commented:
Looks like it's good to merge :+1:  
I just need to fix the Vulkan SDK in CI.

limbonaut commented:
Great stuff! :heart:  Thanks for the contribution.
